onnxruntime/core/providers/cpu/tensor/concat.cc

```diff
@@ -35,66 +35,117 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx,
                                      const std::vector<const Tensor*>& input_tensors,
                                      Prepare& p) const {
   int input_count = static_cast<int>(input_tensors.size());
+
   // Must have atleast one input to concat
   ORT_RETURN_IF_NOT(input_count >= 1, "Must have 1 or more inputs");
 
-  const Tensor* tensor_pointer = input_tensors[0];
-  if (tensor_pointer == nullptr)
-    return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
+  std::vector<int64_t> reference_dims;
+  size_t reference_rank = 0;
+
+  int reference_tensor_index = 0;
+
+  std::vector<int64_t> input_tensor_sizes;
+  input_tensor_sizes.reserve(input_count);
+
+  bool all_inputs_are_empty = true;
+
+  for (int index = 0; index < input_count; ++index) {
+    const auto* input = input_tensors[index];
+    ORT_ENFORCE(input != nullptr, "input count mismatch");
+
+    // find the first tensor that isn't empty
+    // to be used as a reference for all
+    // downstream shape/rank validations of other inputs
+    const auto& shape = input->Shape();
+    const auto num_elements = shape.Size();
+    if (num_elements > 0) {
+      reference_dims = shape.GetDims();
+      reference_rank = reference_dims.size();
+      reference_tensor_index = index;
+      input_tensor_sizes.push_back(num_elements);
+      all_inputs_are_empty = false;
+      break;
+    } else {
+      input_tensor_sizes.push_back(0);
+    }
+  }
 
-  const Tensor& inputs_0 = *tensor_pointer;
-  const auto& inputs_0_dims = inputs_0.Shape().GetDims();
-  const size_t inputs_0_rank = inputs_0_dims.size();
+  if (all_inputs_are_empty) {
+    // Reference dim and reference rank can just come from the first input
+    // No shape/rank validations will be done (as all inputs are empty).
+    // But the rest of the execution flow (filling in the Prepare instance - p)
+    // can use this info.
+    reference_dims = input_tensors[0]->Shape().GetDims();
+    reference_rank = reference_dims.size();
+  }
 
```

Comment:
nit: should we pick the first, or the one with the largest rank? I expect this will never be hit, but our worst-case scenario is a loop with zero iterations and no output shape where we default to rank 1. if say a NonZero was providing an input, whilst it may be empty, it's more likely to have a valid rank. 

And if you were concatentating scalars the rank would end up as 1 anyway, so min would be 1 whatever happens.

Reply:
Yeah it is tricky to pick the right reference tensor in case all of them turn out to be empty. But I prefer to add smarter logic wrt that if we actually hit an issue with that. In most cases, the first one is a reasonably safe choice unless the first input comes from a Loop with zero iterations that didn't have any ONNX inferred shape. It seems like a an extremely rare case. Also, the worst that will happen is we will use that as the reference shape to construct the "empty" output tensor of Concat and some downstream node will hopefully catch the incorrect shape in its validations. All in all, I prefer to wait till we see such an issue before investing in smarter logic for that.